File: frontend/src/types/risk.ts
/** Mirrors backend/app/schemas/risk.py */

/** One multiplier, with the reason it applied.
 *
 * Shown rather than hidden: a reader who can see `× 0.4 test path` can
 * disagree with that specific step. A reader shown only a total cannot.
 */
export interface RiskFactor {
  name: string
  value: number
  reason: string
}

export interface FindingRisk {
  finding_id: number
  score: number
  base: number
  factors: RiskFactor[]
  /** The multiplication written out, e.g. "40 base × 0.8 confidence = 32". */
  explanation: string
  title: string
  severity: string
  file_path: string
  line_start: number
}

export type RiskGrade = 'A' | 'B' | 'C' | 'D' | 'F'

export interface RepositoryRisk {
  repository_id: number
  score: number
  grade: RiskGrade
  /** Bumped when a scoring constant changes; two versions are not comparable. */
  policy_version: number
  finding_count: number
  counts_by_severity: Record<string, number>
  top: FindingRisk[]
}

export interface RiskPoint {
  scan_id: number
  score: number
  grade: string
  policy_version: number | null
  total_findings: number
  finished_at: string
}

export interface RiskHistory {
  repository_id: number
  /** Oldest first — a chart reads left to right. */
  points: RiskPoint[]
}

/** What each grade means in words, so the letter is not just decoration. */
export const GRADE_LABELS: Record<RiskGrade, string> = {
  A: 'Nothing outstanding',
  B: 'Minor issues',
  C: 'Worth addressing',
  D: 'Needs attention',
  F: 'Act on this',
}

File: backend/app/schemas/risk.py
"""Risk request/response schemas.

Every response carries the arithmetic, not just the answer. ``factors`` on each
finding and ``policy_version`` on the total are what let a reader disagree with
a specific step instead of with the number as a whole — which is the only way a
risk model ever gets corrected.
"""

from datetime import datetime

from pydantic import BaseModel


class RiskFactor(BaseModel):
    name: str
    value: float
    reason: str


class FindingRiskRead(BaseModel):
    finding_id: int
    score: float
    base: float
    factors: list[RiskFactor]
    # The multiplication written out, so a tooltip does not have to rebuild it.
    explanation: str
    # Enough of the finding to recognise it without a second request.
    title: str
    severity: str
    file_path: str
    line_start: int


class RepositoryRiskRead(BaseModel):
    repository_id: int
    score: float
    grade: str
    policy_version: int
    # Findings that still count — fixed ones score zero and are excluded.
    finding_count: int
    counts_by_severity: dict[str, int]
    top: list[FindingRiskRead]


class RiskPointRead(BaseModel):
    """One scan's frozen score, for the trend line."""

    scan_id: int
    score: float
    grade: str
    policy_version: int | None
    total_findings: int
    finished_at: datetime


class RiskHistoryRead(BaseModel):
    repository_id: int
    # Oldest first: a chart reads left to right.
    points: list[RiskPointRead]


__all__ = [
    "FindingRiskRead",
    "RepositoryRiskRead",
    "RiskFactor",
    "RiskHistoryRead",
    "RiskPointRead",
]
